plugins/steelseries/fu-steelseries-fizz-hid.c

```diff
@@ -28,19 +28,68 @@ static gboolean
 fu_steelseries_fizz_hid_command(FuDevice *device, guint8 *data, gsize datasz, GError **error)
 {
 	gboolean ret;
-
-	ret = fu_udev_device_pwrite(FU_UDEV_DEVICE(device), 0, data, datasz, error);
-	if (!ret) {
-		g_prefix_error(error, "failed to write report: ");
-		return FALSE;
+	guint8 report_id = 0;
+	guint8 rdata[STEELSERIES_BUFFER_REPORT_SIZE] = {0};
+
+	/* In BT mode the sync and data channels are sharing the device descriptor
+	 * with the management channel. This is the reason why we receive
+	 * "unexpected" packets with 0x01 or 0x05 Report IDs over the same
+	 * descriptor on mouse connecting, waking up or just moving the mouse.
+	 * */
+	while (TRUE) {
+		g_autoptr(GError) pread_error = NULL;
+
+		/* Force the request for each iteration to avoid the
+		 * situation with an infinite loop due the lost single packet.
+		 * This is safe since the device doesn't support the update over BT.
+		 * */
+		ret = fu_udev_device_pwrite(FU_UDEV_DEVICE(device), 0, data, datasz, error);
+		if (!ret) {
+			g_prefix_error(error, "failed to write report: ");
+			return FALSE;
+		}
+
+		/* cleanup the buffer before receiving any data */
+		memset(rdata, 0x00, datasz);
+		ret = fu_udev_device_pread(FU_UDEV_DEVICE(device), 0, rdata, datasz, &pread_error);
+
+		if (!fu_memread_uint8_safe(rdata,
+					   datasz,
+					   STEELSERIES_HID_VERSION_REPORT_ID_OFFSET,
+					   &report_id,
+					   error))
+			return FALSE;
+
+		if (!ret) {
+			/* Since fu_udev_device_pread() treats unexpected data size as error
+			 * we have to check the output additionally since the size of
+			 * unexpected data size from mouse input data is only 16b.
+			 * */
+			if (!g_error_matches(pread_error, G_IO_ERROR, G_IO_ERROR_FAILED) ||
+			    report_id != 0x01) {
+				g_propagate_prefixed_error(error,
+							   g_steal_pointer(&pread_error),
+							   "failed to read report: ");
+				return FALSE;
+			}
+		}
+
+		fu_dump_raw(G_LOG_DOMAIN, "got report", rdata, datasz);
+
+		if (report_id == STEELSERIES_HID_GET_REPORT)
+			break;
```

Comment:
could this loop forever?

Reply:
Only in case if the device doesn't reply with `0x04` report ID for some reason.

Reply:
I think fu_device_retry is going to be much safer